api/core/v1alpha1/device_types_test.go

```diff
@@ -0,0 +1,262 @@
+)
+
+func TestDevice_GetActiveProvisioning(t *testing.T) {
+	now := metav1.Now()
+	oneHourAgo := metav1.NewTime(now.Add(-1 * time.Hour))
+	twoHoursAgo := metav1.NewTime(now.Add(-2 * time.Hour))
+
+	tests := []struct {
+		name    string
+		device  *Device
+		wantNil bool
+	}{
+		{
+			name: "no provisioning entries",
+			device: &Device{
+				Status: DeviceStatus{
+					Provisioning: []ProvisioningInfo{},
+				},
+			},
+			wantNil: true,
+		},
+		{
+			name: "single active provisioning entry",
+			device: &Device{
+				Status: DeviceStatus{
+					Provisioning: []ProvisioningInfo{
+						{
+							StartTime: now,
+							Token:     "token1",
+						},
+					},
+				},
+			},
+			wantNil: false,
+		},
+		{
+			name: "single completed provisioning entry",
+			device: &Device{
+				Status: DeviceStatus{
+					Provisioning: []ProvisioningInfo{
+						{
+							StartTime: twoHoursAgo,
+							EndTime:   oneHourAgo,
+							Token:     "token1",
+						},
+					},
+				},
+			},
+			wantNil: true,
+		},
+		{
+			name: "multiple completed provisioning entries",
+			device: &Device{
+				Status: DeviceStatus{
+					Provisioning: []ProvisioningInfo{
+						{
+							StartTime: metav1.NewTime(now.Add(-3 * time.Hour)),
+							EndTime:   twoHoursAgo,
+							Token:     "token1",
+						},
+						{
+							StartTime: twoHoursAgo,
+							EndTime:   oneHourAgo,
+							Token:     "token2",
+						},
+					},
+				},
+			},
+			wantNil: true,
+		},
+		{
+			name: "active provisioning",
+			device: &Device{
+				Status: DeviceStatus{
+					Provisioning: []ProvisioningInfo{
+						{
+							StartTime: now,
+							Token:     "token-active",
+						},
+						{
+							StartTime: twoHoursAgo,
+							EndTime:   oneHourAgo,
+							Token:     "token-completed",
+						},
+					},
+				},
+			},
+			wantNil: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := tt.device.GetActiveProvisioning()
+			if tt.wantNil {
+				if got != nil {
+					t.Errorf("GetActiveProvisioning() = %v, want nil", got)
+				}
+				return
+			}
+
+			if got == nil {
+				t.Fatalf("GetActiveProvisioning() = nil, want non-nil")
+			}
+
+			want := &tt.device.Status.Provisioning[0]
+			if got.Token != want.Token {
+				t.Errorf("GetActiveProvisioning() token = %v, want %v", got.Token, want.Token)
+			}
+			if !got.StartTime.Equal(&want.StartTime) {
+				t.Errorf("GetActiveProvisioning() startTime = %v, want %v", got.StartTime, want.StartTime)
+			}
+			if !got.EndTime.IsZero() {
+				t.Errorf("GetActiveProvisioning() endTime = %v, want zero", got.EndTime)
+			}
+		})
+	}
+}
+
+func TestDevice_CreateProvisioningEntry(t *testing.T) {
+	now := metav1.Now()
+	oneHourAgo := metav1.NewTime(now.Add(-1 * time.Hour))
+	twoHoursAgo := metav1.NewTime(now.Add(-2 * time.Hour))
+
+	tests := []struct {
+		name        string
+		device      *Device
+		wantErr     bool
+		errContains string
+		validate    func(t *testing.T, device *Device, entry *ProvisioningInfo)
```

Comment:
This validate func feels weird. If validation is that different between test cases, I would simply split them into distinct `Test...` functions, like: 

```go
func TestDevice_CreateProvisioningEntry_Success(t *testing.T) {
  // ...
}

func TestDevice_CreateProvisioningEntry_Success_Completed(t *testing.T) {
  // ...
}

...
```